File: DesignPattern/other_patterns/producer_consumer_pattern.py
import threading
import time
import queue

# Shared queue to store the data
shared_queue = queue.Queue(maxsize=5)  # Set the maximum queue size

# Producer function
def producer(id):
    for i in range(10):
        data = f"Data-{i} from Producer-{id}"
        print(f"Producer-{id} produced: {data}")
        shared_queue.put(data)
        time.sleep(0.1)  # Simulate some work

# Consumer function
def consumer(id):
    while True:
        data = shared_queue.get()
        if data is None:
            break  # Exit the loop when a sentinel value is received
        print(f"Consumer-{id} consumed: {data}")
        shared_queue.task_done()  # Mark the task as done

# Create and start producer threads
producer_threads = []
for i in range(3):
    producer_thread = threading.Thread(target=producer, args=(i,))
    producer_threads.append(producer_thread)
    producer_thread.start()

# Create and start consumer threads
consumer_threads = []
for i in range(2):
    consumer_thread = threading.Thread(target=consumer, args=(i,))
    consumer_threads.append(consumer_thread)
    consumer_thread.start()

# Wait for all producer threads to finish
for producer_thread in producer_threads:
    producer_thread.join()

# Send sentinel values to indicate the end of data
for _ in range(len(consumer_threads)):
    shared_queue.put(None)

# Wait for all consumer threads to finish
for consumer_thread in consumer_threads:
    consumer_thread.join()

print("All threads have finished.")



'''
- The Producer-Consumer pattern is a classic multi-threading synchronization pattern where one or more producer threads produce data, and one or more 
consumer threads consume the data. In Python, you can implement the Producer-Consumer pattern using the threading module for thread management and a 
shared queue for communication between producers and consumers. Here's a basic example:

- In this Python example: We use the queue.Queue class to create a shared queue with a maximum size of 5. The queue is used to pass data between producers and consumers.

- The producer function simulates a producer by producing data and putting it into the shared queue. Producers are designed to produce data continuously for demonstration purposes.

- The consumer function simulates a consumer by consuming data from the shared queue. Consumers keep consuming data until they receive a sentinel 
value (None), indicating the end of data.

- We create and start three producer threads and two consumer threads. The producers and consumers operate concurrently.

- After all producer threads have finished, we send sentinel values (None) to indicate the end of data for consumers.

- The code waits for all consumer threads to finish and prints a message when all threads have finished.

- This example demonstrates a basic Producer-Consumer pattern in Python using threads and a shared queue for synchronization and communication between producers and consumers.
'''

File: DesignPattern/other_patterns/producer_consumer_pattern.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <condition_variable>

const int buffer_size = 5;  // Size of the shared buffer
std::vector<int> buffer;    // Shared buffer to store data
std::mutex mtx;             // Mutex to protect access to the buffer
std::condition_variable buffer_not_full;  // Signaled when the buffer is not full
std::condition_variable buffer_not_empty; // Signaled when the buffer is not empty

void producer(int id) {
    for (int i = 0; i < 10; ++i) {
        std::unique_lock<std::mutex> lock(mtx);
        while (buffer.size() >= buffer_size) {
            // Buffer is full; wait for a consumer to make space
            buffer_not_full.wait(lock);
        }
        int data = i * 100 + id;
        buffer.push_back(data);
        std::cout << "Producer " << id << " produced: " << data << std::endl;
        buffer_not_empty.notify_all(); // Notify consumers that the buffer is not empty
    }
}

void consumer(int id) {
    for (int i = 0; i < 10; ++i) {
        std::unique_lock<std::mutex> lock(mtx);
        while (buffer.empty()) {
            // Buffer is empty; wait for a producer to add data
            buffer_not_empty.wait(lock);
        }
        int data = buffer.back();
        buffer.pop_back();
        std::cout << "Consumer " << id << " consumed: " << data << std::endl;
        buffer_not_full.notify_all(); // Notify producers that the buffer is not full
    }
}

int main() {
    std::vector<std::thread> producers;
    std::vector<std::thread> consumers;

    for (int i = 1; i <= 3; ++i) {
        producers.emplace_back(producer, i);
        consumers.emplace_back(consumer, i);
    }

    for (auto& producer_thread : producers) {
        producer_thread.join();
    }

    for (auto& consumer_thread : consumers) {
        consumer_thread.join();
    }

    return 0;
}


/*
- The Producer-Consumer pattern is a classic multi-threading synchronization pattern where one or more producer threads produce data, and one or more 
consumer threads consume the data. The data is typically stored in a shared buffer. In this example, I'll provide a basic implementation of the 
Producer-Consumer pattern in C++ using std::thread and std::mutex.

- In this example: We have a shared buffer of a fixed size (buffer_size) to store data produced by producers and consumed by consumers.

- The producer function simulates a producer, and the consumer function simulates a consumer. Producers generate data and push it into the buffer, 
while consumers pop data from the buffer. They use condition variables to synchronize access to the buffer.

- The main function creates and starts producer and consumer threads. In this example, we have three producers and three consumers.

- Producers produce data and notify consumers when the buffer is not empty. Consumers consume data and notify producers when the buffer is not full.

- This example demonstrates a basic Producer-Consumer pattern in C++ using threads, mutexes, and condition variables to ensure safe and synchronized access to a shared buffer.
*/